### PR TITLE
Cleanup of proj file

### PR DESCRIPTION
**Pull request by martijn00**
This is to include the right new metadata into the nuget package, and do some general maintenance.

### CODE REVIEW

File: ReGitLint/Cleanup.cs [hunk 1]
```diff
@@ -369,15 +366,15 @@ private void PrintFixCommand()
     {
         var args = string.Join(" ", Environment.GetCommandLineArgs().Skip(1));
 
-        if (args.Contains("--jenkins"))
+        if (args.Contains("--jenkins", StringComparison.OrdinalIgnoreCase))
```
Comment (bkoelman) [inline ReGitLint/Cleanup.cs]: This looks wrong; command line switches are case-sensitive.

File: icon.png
Comment (bkoelman) [inline icon.png:1]: I don't think it's a good idea to copy another company's logo into an open-source project. Have you considered the legal implications?

File: ReGitLint/Cleanup.cs [hunk 2]
```diff
@@ -648,7 +645,7 @@ private int RunCleanupCode(string include, string slnFile)
 
         void ErrorCallback(string data)
         {
-            if (data.Contains("No items were found to cleanup"))
+            if (data.Contains("No items were found to cleanup", StringComparison.OrdinalIgnoreCase))
```
Comment (bkoelman) [inline ReGitLint/Cleanup.cs]: This should match the exact message, not ignore casing.

File: Directory.Build.props [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        <PackageLicenseExpression>MIT</PackageLicenseExpression>
+        <PackageProjectUrl>https://github.com/sethreno/ReGitLint</PackageProjectUrl>
+        <Authors>Seth Reno</Authors>
+        <Owners>Seth Reno</Owners>
+        <PackageTags>.net;c#;git;lint;ci</PackageTags>
+        <PackageReleaseNotes>https://github.com/sethreno/ReGitLint/releases</PackageReleaseNotes>
+        <PackageReadmeFile>readme.md</PackageReadmeFile>
+        <PackageRequireLicenseAcceptance>false</PackageRequireLicenseAcceptance>
+        <NeutralLanguage>en-US</NeutralLanguage>
+        <MinClientVersion>2.14</MinClientVersion>
```
Comment (bkoelman) [inline Directory.Build.props:13]: I don't see the point of moving these into a separate file, when we know there'll always be one project to publish. Extracting these just makes it harder to find these settings.

File: Directory.Build.props [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    </PropertyGroup>
+
+    <ItemGroup>
+        <None Include="$(MSBuildThisFileDirectory)\LICENSE" Pack="true" PackagePath="\" />
+        <None Include="$(MSBuildThisFileDirectory)\icon.png" Pack="true" PackagePath="\" />
+        <None Include="$(MSBuildThisFileDirectory)\readme.md" Pack="true" PackagePath="\" />
+    </ItemGroup>
+
+    <ItemGroup>
+        <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.1.1" PrivateAssets="All" />
```
Comment (bkoelman) [inline Directory.Build.props]: I don't see the point of adding SourceLink and related properties. This is a CLI tool, not a regular NuGet package that developers can step into while debugging.

File: Directory.Build.props [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+        <DesignTimeBuild>false</DesignTimeBuild>
+        <DebugType>portable</DebugType>
+        <DebugSymbols>true</DebugSymbols>
+
+        <GenerateDocumentationFile>true</GenerateDocumentationFile>
+        <GeneratePackageOnBuild Condition=" '$(IsTestProject)' != 'true'">true</GeneratePackageOnBuild>
+        <PackageIcon>icon.png</PackageIcon>
+        <ContinuousIntegrationBuild>true</ContinuousIntegrationBuild>
+        <Deterministic>true</Deterministic>
```
Comment (bkoelman) [inline Directory.Build.props:47]: I don't consider these an improvement. It just adds unneeded complexity that isn't tailored to this project. For example, deterministic builds are enabled by default in .NET SDK projects.

File: ReGitLint.sln [hunk 6]
```diff
@@ -1,22 +1,32 @@
...
 		{A92B8CE9-A19D-4BA8-AB20-39B55890DB6C}.Debug|Any CPU.Build.0 = Debug|Any CPU
 		{A92B8CE9-A19D-4BA8-AB20-39B55890DB6C}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{A92B8CE9-A19D-4BA8-AB20-39B55890DB6C}.Release|Any CPU.Build.0 = Release|Any CPU
 	EndGlobalSection
+	GlobalSection(SolutionProperties) = preSolution
+		HideSolutionNode = FALSE
+	EndGlobalSection
+	GlobalSection(ExtensibilityGlobals) = postSolution
+		SolutionGuid = {5C07519F-D58A-4BE5-81DB-D06FD4F50B3A}
+	EndGlobalSection
```
Comment (bkoelman) [inline ReGitLint.sln:31]: Why was this added?

File: ReGitLint/Cleanup.cs [hunk 7]
```diff
@@ -340,7 +337,7 @@ private List<string> GetIncludeArgs(HashSet<string> filePaths)
                 // jb codecleanup requires file paths relative to the sln
                 var jbFilePath = Path.GetRelativePath(slnDirAbs, filePathAbs);
 
-                if (jbFilePath.StartsWith(".."))
+                if (jbFilePath.StartsWith("..", StringComparison.OrdinalIgnoreCase))
```
Comment (bkoelman) [inline ReGitLint/Cleanup.cs]: There's nothing in ".." that should consider case-insensitivity.

File: ReGitLint/Cleanup.cs [hunk 8]
```diff
@@ -469,7 +466,7 @@ string commitB
                 commitA = commitB;
             if (string.IsNullOrEmpty(commitB))
                 commitB = commitA;
-            if (commitA == commitB)
+            if (string.Equals(commitA, commitB, StringComparison.OrdinalIgnoreCase))
```
Comment (bkoelman) [inline ReGitLint/Cleanup.cs]: If these can ever only be commit hashes, the change is fine. However, a case-sensitive comparison is required if this could also be a branch name.

File: ReGitLint/Cleanup.cs [hunk 9]
```diff
@@ -487,7 +484,7 @@ private static bool DoesCommitExist(string sha)
 
         void OutputCallback(string data)
         {
-            if (data.StartsWith("commit"))
+            if (data.StartsWith("commit", StringComparison.OrdinalIgnoreCase))
```
Comment (bkoelman) [inline ReGitLint/Cleanup.cs]: This captures output, so should not ignore case.

File: Directory.Build.props [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    </PropertyGroup>
+
+    <PropertyGroup Condition=" '$(Configuration)' == 'Debug' ">
+        <DebugType>full</DebugType>
+        <DebugSymbols>true</DebugSymbols>
+    </PropertyGroup>
+
+    <ItemGroup>
+        <None Include="$(MSBuildThisFileDirectory)\LICENSE" Pack="true" PackagePath="\" />
```
Comment (bkoelman) [inline Directory.Build.props:56]: Packing this increases the NuGet download size for no reason. Using `PackageLicenseExpression` is more efficient.

File: Directory.Build.props [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        <GeneratePackageOnBuild Condition=" '$(IsTestProject)' != 'true'">true</GeneratePackageOnBuild>
+        <PackageIcon>icon.png</PackageIcon>
+        <ContinuousIntegrationBuild>true</ContinuousIntegrationBuild>
+        <Deterministic>true</Deterministic>
+    </PropertyGroup>
+
+    <PropertyGroup Condition=" '$(Configuration)' == 'Debug' ">
+        <DebugType>full</DebugType>
+        <DebugSymbols>true</DebugSymbols>
+    </PropertyGroup>
```
Comment (bkoelman) [inline Directory.Build.props:53]: Isn't this the default behavior?

File: Directory.Build.props [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        <NeutralLanguage>en-US</NeutralLanguage>
+        <MinClientVersion>2.14</MinClientVersion>
+
+        <PublishRepositoryUrl>true</PublishRepositoryUrl>
+        <RepositoryUrl>https://github.com/sethreno/ReGitLint</RepositoryUrl>
+        <RepositoryType>git</RepositoryType>
+        <Product>$(AssemblyName) ($(TargetFramework))</Product>
+        <Version>6.1.0-alpha</Version>
+
+        <!--<Nullable>enable</Nullable>-->
```
Comment (bkoelman) [inline Directory.Build.props]: Why was this added in commented-out form?

File: Directory.Build.props [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        <PackageReadmeFile>readme.md</PackageReadmeFile>
+        <PackageRequireLicenseAcceptance>false</PackageRequireLicenseAcceptance>
+        <NeutralLanguage>en-US</NeutralLanguage>
+        <MinClientVersion>2.14</MinClientVersion>
+
+        <PublishRepositoryUrl>true</PublishRepositoryUrl>
+        <RepositoryUrl>https://github.com/sethreno/ReGitLint</RepositoryUrl>
+        <RepositoryType>git</RepositoryType>
+        <Product>$(AssemblyName) ($(TargetFramework))</Product>
+        <Version>6.1.0-alpha</Version>
```
Comment (bkoelman) [inline Directory.Build.props]: This used to be `PackageVersion`. Changing this possibly breaks the existing auto-versioning logic, which I'm not very familiar with.

File: ReGitLint/ReGitLint.csproj [hunk 14]
```diff
@@ -1,23 +1,14 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFrameworks>netcoreapp3.1;net5.0;net6.0</TargetFrameworks>
-    <LangVersion>10.0</LangVersion>
-    <ImplicitUsings>enable</ImplicitUsings>
+    <TargetFrameworks>netcoreapp3.1;net5.0;net6.0;net7.0</TargetFrameworks>
```
Comment (bkoelman) [inline ReGitLint/ReGitLint.csproj]: What's the point of adding net70 as a target? It increases the NuGet download size, while net60 is LTS and still fully supported (and installed by Visual Studio out of the box).

**Review by bkoelman — CHANGES_REQUESTED** [review]
See comments.

File: Directory.Build.props (hunk 3 above)
Comment (martijn00) [reply Directory.Build.props:13]: This is standard in all newer .NET project structures. It makes it easier for other devs to use.

File: Directory.Build.props (hunk 5 above)
Comment (martijn00) [reply Directory.Build.props:47]: To my knowledge Deterministic is not enabled by default. I think these will improve the visibility on the nuget website. Can you specify what you would like to remove?

File: ReGitLint.sln (hunk 6 above)
Comment (martijn00) [reply ReGitLint.sln:31]: This is the new file added

File: Directory.Build.props (hunk 10 above)
Comment (martijn00) [reply Directory.Build.props:56]: PackageLicenseExpression alone is not enough. If you don't include the actual license in the nuget it doesn't cover it.

File: Directory.Build.props (hunk 11 above)
Comment (martijn00) [reply Directory.Build.props:53]: Not to my knowledge. I use this for all my projects.

File: Directory.Build.props (hunk 3 above)
Comment (bkoelman) [reply Directory.Build.props:13]: Why do you believe that's the case? If I follow the steps at https://learn.microsoft.com/en-us/dotnet/core/tools/global-tools-how-to-create and run:
```
dotnet new console -n mytool
```
it doesn't generate a `Directory.builds.props` file.

File: Directory.Build.props (hunk 5 above)
Comment (bkoelman) [reply Directory.Build.props:47]: It is, according to https://devblogs.microsoft.com/dotnet/producing-packages-with-source-link/:

> While deterministic builds are enabled by default in .NET SDK projects, there is an extra property, ContinuousIntegrationBuild, to set on the build server to normalize stored file paths. These should not be enabled during local dev or the debugger won’t be able to find the local source files.

Please remove all of them, they add no value in this case.

File: ReGitLint.sln (hunk 6 above)
Comment (bkoelman) [reply ReGitLint.sln:31]: When I run:
```
dotnet new sln
```
this doesn't get added.

File: Directory.Build.props (hunk 10 above)
Comment (bkoelman) [reply Directory.Build.props:56]: According to https://learn.microsoft.com/en-us/nuget/reference/nuspec#license, it is. Adding the file is only needed when a non-standard license is used.

File: Directory.Build.props (hunk 11 above)
Comment (bkoelman) [reply Directory.Build.props:53]: DebugType: see https://stackoverflow.com/questions/46902377/what-is-the-difference-between-using-debugtypefull-debugtype-and-debugtype:
> The difference is that the "full" type emits a classic windows PDB symbol file which is complex and poorly documented. The "portable" PDB format is a new open-source format that can be created and used on all platforms.

By not explicitly specifying these, the proper defaults are used, depending on the project type, configuration, target runtime etc. The only case I'm aware of to force "full", is to make some code coverage tools work, which does not apply here.

DebugSymbols: I'm pretty sure these are on for debug builds. Otherwise you wouldn't get .pdb files.

File: Directory.Build.props [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+        <MinClientVersion>2.14</MinClientVersion>
+
+        <PublishRepositoryUrl>true</PublishRepositoryUrl>
+        <RepositoryUrl>https://github.com/sethreno/ReGitLint</RepositoryUrl>
+        <RepositoryType>git</RepositoryType>
+        <Product>$(AssemblyName) ($(TargetFramework))</Product>
+
+        <Nullable>enable</Nullable>
+        <ImplicitUsings>enable</ImplicitUsings>
+        <LangVersion>latest</LangVersion>
```
Comment (bkoelman) [inline Directory.Build.props:22]: It's not recommended to set this, according to https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version:
> The latest C# compiler determines a default language version based on your project's target framework or frameworks. Visual Studio doesn't provide a UI to change the value, but you can change it by editing the csproj file. The choice of default ensures that you use the latest language version compatible with your target framework. You benefit from access to the latest language features compatible with your project's target. This default choice also ensures you don't use a language that requires types or runtime behavior not available in your target framework. Choosing a language version newer than the default can cause hard to diagnose compile-time and runtime errors.

File: Directory.Build.props [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+        <AnalysisMode>Default</AnalysisMode>
+
+        <IsTestProject>$(MSBuildProjectName.Contains('Test'))</IsTestProject>
+    </PropertyGroup>
+
+    <PropertyGroup Condition=" '$(Configuration)' == 'Release' ">
+        <!--<SymbolPackageFormat>snupkg</SymbolPackageFormat>-->
+        <IncludeSymbols>true</IncludeSymbols>
+        <AllowedOutputExtensionsInPackageBuildOutputFolder>$(AllowedOutputExtensionsInPackageBuildOutputFolder);.pdb</AllowedOutputExtensionsInPackageBuildOutputFolder>
+        <EmbedUntrackedSources>true</EmbedUntrackedSources>
```
Comment (bkoelman) [inline Directory.Build.props:37]: This is related to SourceLink and should be removed as well.

File: ReGitLint/Cleanup.cs [hunk 17]
```diff
@@ -469,7 +466,7 @@ string commitB
                 commitA = commitB;
             if (string.IsNullOrEmpty(commitB))
                 commitB = commitA;
-            if (commitA == commitB)
+            if (string.Equals(commitA, commitB, StringComparison.Ordinal))
```
Comment (bkoelman) [inline ReGitLint/Cleanup.cs:469]: Please revert this change, `string.operator==` already performs an ordinal comparison.

File: ReGitLint/ReGitLint.csproj [hunk 18]
```diff
@@ -2,22 +2,14 @@
...
     <ToolCommandName>regitlint</ToolCommandName>
-    <PackageOutputPath>./nupkg</PackageOutputPath>
-    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
-    <RepositoryUrl>https://github.com/sethreno/ReGitLint</RepositoryUrl>
-    <Authors>Seth Reno</Authors>
-    <PackageDescription>Low friction code formatting for .net projects</PackageDescription>
-    <Copyright>Copyright 2020 Seth Reno</Copyright>
-    <PackageTags>.net;c#;git;lint;ci</PackageTags>
-    <PackageLicenseExpression>MIT</PackageLicenseExpression>
+    <PackageOutputPath>$(SolutionDir)/nupkg</PackageOutputPath>
```
Comment (bkoelman) [inline ReGitLint/ReGitLint.csproj:12]: What are the implications of changing this? Will NuGet auto-publish still work if binaries are written to `/nupkg` instead of `/ReGitLint/nupkg`? I don't see the point of changing this, risking things to break.

File: icon.png
Comment (bkoelman) [inline icon.png:1]: Why do we need an icon? Having an icon is not a goal in itself; it's supposed to make people visually recognize they're looking at ReGitLint.

Given that we don't have one today, I believe it's better to leave it out than to place something generic here.

**Review by bkoelman — COMMENTED** [review]
I appreciate the work you've put into this PR, but I'm starting to wonder if it is worth pursuing. It would be great to tweak things where we have good reasons to, but I get the impression this is just a bunch of defaults being thrown in, combined with unneeded things.

ReGitLint is a very simple (script-like) command-line tool that just does its job, not a low-level library or an enterprise application. So I favor keeping things simple and using implicit defaults unless we have good reasons to deviate from that.

The ordinal comparisons make sense on hot code paths for large collections because culture-sensitive comparisons are slightly more expensive (in the order of microseconds). But the thing is: this is a command-line tool, not a low-level framework library, so it doesn't really matter.

**Comment by bkoelman** [comment]
I'm closing this, as the work seems to be abandoned. Please let me know if you're still working on this, and I'll reopen it.
